src/actions/attack.cpp

```diff
@@ -1561,6 +1573,157 @@ std::pair<int, map_location> under_leadership(const unit_map& units, const map_l
 	return abil.highest("value");
 }
 
+//begin of weapon emulates function.
+
+bool unit::abilities_filter_matches(const config& cfg, bool attacker, int res) const
+{
+	if(!(cfg["active_on"].empty() || (attacker && cfg["active_on"] == "offense") || (!attacker && cfg["active_on"] == "defense"))) {
+		return false;
+	}
+
+	if(!unit_abilities::filter_base_matches(cfg, res)) {
+		return false;
+	}
+
+	return true;
+}
+
+//functions for emulate weapon specials.
+//filter opponent and affect self/opponent/both option.
+bool unit::ability_filter_opponent(const std::string& ability,const config& cfg,const map_location& loc) const
+{
+	const config &filter = cfg.child("filter_opponent");
+	if (!filter) return true;
+	return unit_filter(vconfig(filter)).set_use_flat_tod(ability == "illuminates").matches(*this, loc);
+}
+
+bool leadership_affects_self(const std::string& ability,const unit_map& units, const map_location& loc, const_attack_ptr weapon,const_attack_ptr opp_weapon)
+{
+	const unit_map::const_iterator un = units.find(loc);
+	if(un == units.end()) {
+		return false;
+	}
+
+     unit_ability_list abil = un->get_abilities(ability, weapon, opp_weapon);
+     for(unit_ability_list::iterator i = abil.begin(); i != abil.end();) {
+              const std::string& apply_to = (*i->first)["apply_to"];
+            if(!(apply_to.empty()||apply_to == "both"||apply_to == "under")) {
+                return false;
+     } else {
+         return true;
+         }
+         ++i;
+         }
+     return false;
+}
+
+bool leadership_affects_opponent(const std::string& ability,const unit_map& units, const map_location& loc, const_attack_ptr weapon,const_attack_ptr opp_weapon)
+{
+	const unit_map::const_iterator un = units.find(loc);
+	if(un == units.end()) {
+		return false;
+	}
+
+     unit_ability_list abil = un->get_abilities(ability, weapon, opp_weapon);
+     for(unit_ability_list::iterator i = abil.begin(); i != abil.end();) {
+              const std::string& apply_to = (*i->first)["apply_to"];
+            if(apply_to.empty()||!(apply_to == "both"||apply_to == "opponent")) {
+                return false;
+     } else {
+         return true;
+         }
+         ++i;
+         }
+     return false;
```

Comment:
Can only return false if the is nothing in the list you're iterating over.

Reply:
By default leadership_affect_self must be true and affect_opponent must return false.